frontend/src/liboperator/operator/operator.go

```diff
@@ -169,6 +174,38 @@ func deviceEndpoint(c *JSONUnix, pool *DevicePool, config apiv1.InfraConfig) {
 	}
 }
 
+func openwrt(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
+	vars := mux.Vars(r)
+	devId := vars["deviceId"]
+	dev := pool.GetDevice(devId)
+	if dev == nil {
+		http.Error(w, "Device not found", http.StatusNotFound)
+		return
+	}
+
+	openwrtDevId, ok := dev.info.(map[string]interface{})["openwrt_device_id"]
+	if !ok {
+		http.Error(w, "Device obtaining Openwrt not found", http.StatusNotFound)
+		return
+	}
+
+	path := vars["path"]
+	if devId == openwrtDevId {
+		openwrtAddr, ok := dev.info.(map[string]interface{})["openwrt_addr"]
+		if !ok {
+			http.Error(w, "Openwrt address not found", http.StatusNotFound)
+			return
+		}
+		url, _ := url.Parse("http://" + openwrtAddr.(string))
+		proxy := httputil.NewSingleHostReverseProxy(url)
+		r.URL.Path = "/devices/" + devId + "/openwrt" + path
+		proxy.ServeHTTP(w, r)
+	} else {
+		http.Redirect(w, r, "/devices/" + openwrtDevId.(string) + "/openwrt" + path, http.StatusFound)
```

Comment:
if `devId` is not the same as `openwrtDevId` wouldn't `pool.GetDevice(devId)`(line 180) fail to find the device when this redirect is called?

It seems to me the correct approach would be to not redirect and always use `openwrtDevId` instead of `devId` when building the URL to proxy.

Reply:
Originally openwrt luci interface has some paths like `/cgi-bin/luci`, `/luci-static/*`, and `/ubus`. While accessing to luci interface, it tries to access and redirects to each other. It's not an issue when accessing openwrt IP address itself like `https://192.168.94.2`, but it failes to access some paths via redirection while using reverse proxy. For example, when I access `https://localhost:1443/devices/cvd-1/openwrt/cgi-bin/luci` then luci interfaces tries to access `https://localhost:1443/luci-static` but that path is not for that openwrt luci interface.

So I applied a change to AOSP adding prefix for all paths in luci interface, like modifying `/luci-static` into `/devices/cvd-1/openwrt/luci-static`. However, it still have an issue when running multiple CF instances with only one openwrt instance such as `cvd start -instance_nums=1,2`. In this case, the openwrt luci interface uses the path `/devices/cvd-1/openwrt/*` That's the reason I applied redirection from `/devices/cvd-2/openwrt/*` into `/devices/cvd-1/openwrt/*` because the actual openwrt luci interface runs with the path `/devices/cvd-1/openwrt/*`.

Not using redirection in operator.go still works so I applied your suggestions, but functionally it still redirects to the path based on device id obtaining openwrt instance because of the redirection within luci interface itself.

Reply:
Sure! The related AOSP change is aosp/2585253. When OpenWRT instance is launched with crosvm, the instance name like `cvd-1` is delivered via `--params` argument of crosvm. I'll add you as cc of that change.